### DISCUSSION TITLE
add_all not using batching

### DISCUSSION THREAD

**Discussion by SnoozeFreddo** [sqlalchemy/sqlalchemy#11899] [Usage Questions] [answered]
Hello there,

i am using postgresl + asyncpg.

When i try to use session.add_all(MyDbModelList) and db.commit() i get 300 sql executions instead of 1. Asyncpg   has executemany so it should work, but it isn't.

"name": "sqlalchemy.engine.Engine", "message": "[insertmanyvalues 209/300 (ordered; batch not supported)]

i could use 

db.execute(insert(MyDbModel), list_of_dict_rows)

However there is a Parent/Child (One2Many) relation and it's just ignoring all the child values for that field. 

Any idea what i am doing wrong?

**Comment by zzzeek** [comment]
`add_all()` does not imply anything about what kind of INSERT statement is emitted, only that you want all those objects to be inserted on the next flush.

the logging indicates insertmanyvalues is in use but it's not able to use batching for these statements.   if you are using a server side UUID for primary key or something like that, then batching might not be available in ORM "ordered" mode.   [using bulk insert](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements) however won't require ordering and you can use that.

**Comment by SnoozeFreddo** [comment]
I tried to use the bulk insert statement before. That's when the relations(Childs) would get ignored and not created along the Parent. 

I also do use server side UUID sadly.

**Reply by zzzeek** [reply]
you can have performance, ease of use, or not use autoincrementing integer primary keys.   choose any two :)

**Reply by SnoozeFreddo** [reply]
I choose performance and not use autoincrement. 

But how do i execute performance with one2many? 

Batch_Create the Parents and with a second query bulk_add the "many" side with parent_id set based on the returned objects from the batch_create?  What's the best practice?

**Reply by zzzeek** [reply]
batching will also work if you use a client side uuid generator instead of server side.  set column `default=uuid.uuid4` and then they will be generated ahead of time with batching (should be).

**Reply by zzzeek** [reply, marked as answer]
confirmed this works even if you have a server side default on other columns, I think this is your easiest approach

```py
from __future__ import annotations

import datetime
import uuid

from sqlalchemy import create_engine
from sqlalchemy import func
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        # will batch
        default=uuid.uuid4,
        # will not batch
        # server_default=func.gen_random_uuid(),
    )
    data: Mapped[str]

    # server defaults that are not PK still eagerly fetched
    create_date: Mapped[datetime.datetime] = mapped_column(
        server_default=func.now()
    )


e = create_engine("postgresql+psycopg://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(data=f"{i}") for i in range(50)])
s.commit()

```
